### DISCUSSION TITLE
Unexpected 404 Response Despite Correct Terminal Output Logging

### DISCUSSION THREAD

**Discussion by bajrangCoder** [honojs/hono#1891] [Q&A] [answered]
I recently transitioned from `expressjs` to `hono`, and I've encountered an issue where the server returns a `404 Not Found` response instead of the expected `output`. Strangely, the terminal logs the correct output. Previously, I encountered an error about content not being finalized. After consulting the documentation, I discovered the use of `await next()`. However, since implementing this, I've been facing a 404 error. As I'm new to hono, I would appreciate any assistance with resolving this problem.

**Code for that endpoint**
```ts
app.post("/execute-command", async (c, next) => {
    try {
      const { command } = await c.req.json();
      if (!command) {
        return c.json({ error: "Command is required." }, 400);
      }

      // Execute the command using node-pty
      const term = pty.spawn(
        process.platform === "win32" ? "cmd.exe" : "bash",
        ["-c", command],
        {
          name: "xterm-256color",
          cols: 80,
          rows: 24,
          cwd: process.platform === "win32" ? undefined : process.env.HOME,
        }
      );

      const pattern = [
        "[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]+)*|[a-zA-Z\\d]+(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)",
        "(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-nq-uy=><~]))",
      ].join("|");
      const ansiRegex = new RegExp(pattern, undefined);
      const ansiRegex2 = new RegExp(pattern, "g");

      let output = "";

      // Listen for data events (output from the command)
      term.onData((data) => {
        output += data;
      });

      // Listen for the process to exit
      term.onExit(() => {
        console.log(output: ansiRegex.test(output)
            ? output.replace(ansiRegex2, "")
            : output)
        // Send the parsed output back to the client
        return c.json({
          output: ansiRegex.test(output)
            ? output.replace(ansiRegex2, "")
            : output,
        });
      });
      await next()
    } catch (error) {
      console.error(error);
      return c.json({ error: "Failed to execute command" }, 500);
    }
  });
```

**Comment by EdamAme-x** [comment]
What is the result you want when `await next()` is just execute?

**Reply by EdamAme-x** [reply]
If you want to return a response after the terminal process has terminated.

```js
return c.json({
          output: ansiRegex.test(output)
            ? output.replace(ansiRegex2, "")
            : output,
 });
 ```
 
 I think this part is problematic because it is only done as the return value of the callback function.

**Reply by bajrangCoder** [reply]
Finally, Thanks for response 😃

**Reply by bajrangCoder** [reply]
What can I do in this situation?

**Reply by EdamAme-x** [reply]
Is the problem likely to be resolved?

**Reply by EdamAme-x** [reply]
ok
i thinking now...

**Reply by EdamAme-x** [reply]
sry
Is `term.onExit` an async function?

**Reply by bajrangCoder** [reply]
I guess no
https://github.com/microsoft/node-pty/blob/d6ce76a801f68730f7204b7b27bb21dbf8e63001/typings/node-pty.d.ts#L147

**Reply by EdamAme-x** [reply]
If asynchronous, use `await`, declare a variable for the response in the scope of the handler, and assign it in the callback.
Ideally, finally the response variable should be returned.

**Reply by EdamAme-x** [reply]
Sounds like a synchronous function.
Then I'd just do the same thing without the `await`.

**Reply by bajrangCoder** [reply]
Ok got it 👍
Thanks 🙏

**Reply by EdamAme-x** [reply]
ty

**Comment by bajrangCoder** [comment]
> If asynchronous, use `await`, declare a variable for the response in the scope of the handler, and assign it in the callback. Ideally, finally the response variable should be returned.

I have tried how you have mentioned but it returns the dispose function instead of the output

**Reply by EdamAme-x** [reply]
Perhaps you misunderstood the intent.
I came up with a solution.

How about wrapping it in a Promise object and resolving it when onExit is executed?

**Reply by EdamAme-x** [reply, marked as answer]
new Promise((resolve, reject) => {
    ...
    term.onExit(() => {
         ...
         resolve(result)
    })
})

**Reply by bajrangCoder** [reply]
yeah, this promise method worked, Thanks
